neptune/codelistsdataset.go

```diff
@@ -0,0 +1,155 @@
+satisfy all of:
+    1) code lists that match the requested code list ID.
+    2) code lists of the requested edition.
+    3) codes that match the requested code value.
+    4) datasets that are related to qualifying codes by *inDataset* edges.
+    5) datasets that have the *isPublished* state true.
+
+Each such result from the database (potentially) has the properties:
+    - dimensionName (what the dataset calls this dimension)
+    - datasetEdition
+    - version
+
+The results however include all permuations of dimensionName and 
+datasetEdition - BUT ONLY CITES the most recent dataset *version* of those 
+found for that permuation.
+
+*/
+func (n *NeptuneDB) GetCodeDatasets(ctx context.Context, codeListID, edition string, code string) (*models.Datasets, error) {
+
+	// Emit the query and parse the responses.
+	qry := fmt.Sprintf(query.GetCodeDatasets, codeListID, edition, code)
+	responses, err := n.getStringList(qry)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Gremlin GetCodeDatasets failed: %q", qry)
+	}
+
+	// Isolate the individual records from the flattened response.
+	// [['dim', 'edition', 'version'], ['dim', 'edition', ...]]
+	responseTriples, err := createTriples(responses)
+	if err != nil {
+		return nil, errors.Wrap(err, "Cannot create triples.")
+	}
+
+	// Build datastructure to capture only latest dataset versions.
+	dimensionNameToEditions, err := buildDim2Edition(responseTriples)
+	if err != nil {
+		return nil, errors.Wrap(err, "Cannot isolate latest versions.")
+	}
+
+	// Package up the model-ised response.
+	response := buildResponse(dimensionNameToEditions, code, codeListID)
+	return response, nil
+}
+
+/*
+createTriples splits a list of strings into clumps of 3
+*/
+func createTriples(responses []string) ([][]string, error) {
+	var responseTriples = [][]string{}
+	const stride = 3 // I.e. dimesionName, edition, version
+	if len(responses)%stride != 0 {
+		return nil, errors.New("List length is not divisible by 3")
+	}
+	nInstances := len(responses) / stride
+	for i := 0; i < nInstances; i++ {
+		offset := i * stride
+		dimensionName := responses[offset+0]
+		dataSetEdition := responses[offset+1]
+		versionStr := responses[offset+2]
+		responseTriples = append(responseTriples, []string{dimensionName, dataSetEdition, versionStr})
+	}
+	return responseTriples, nil
+}
+
+// These (nested) maps track the latest version cited by any combination
+// of dimensionName and dataset edition.
+
+type editionToLatestVersion map[string]int
+type dim2Edition map[string]editionToLatestVersion
+
+/*
+buildDim2Edition consumes a list of triples such as
+["dimName1", "datasetEdition1", "version4"], and builds a dim2Edition
+structure based on the latest versions available for each combination of
+dimension name and dataset edition.
+*/
+func buildDim2Edition(responseTriples [][]string) (dim2Edition, error) {
+	d2e := dim2Edition{}
+	for _, respTriple := range responseTriples {
+		dimensionName := respTriple[0]
```

Comment:
(lost this comment in the git-wilderness, so apologies if this ends up as a duplicate)

Unfortunately, nesting under dimensionName won't be specific enough for our case, we need to be including the Dataset ID as a differentiator.

https://api.beta.ons.gov.uk/v1/code-lists/admin-geography/editions/one-off/codes/K02000001/datasets
If you consider this endpoint, the dimesnion Name is the same across all of the subdocuments, and their editions labels are largely the same as well, but there is a separate document for every dataset ID.

We'll need to include projecting the dataset_ID field in the query, and expand these helpers to also separate results based on it.

Reply:
fixed and pushed.